lib/icinga/checkable-check.cpp

```diff
@@ -494,12 +494,22 @@ void Checkable::ProcessCheckResult(const CheckResult::Ptr& cr, const MessageOrig
 		int suppressed_types_before (GetSuppressedNotifications());
 		int suppressed_types_after (suppressed_types_before | suppressed_types);
 
-		for (int conflict : {NotificationProblem | NotificationRecovery, NotificationFlappingStart | NotificationFlappingEnd}) {
-			/* E.g. problem and recovery notifications neutralize each other. */
+		int conflict = NotificationFlappingStart | NotificationFlappingEnd;
+		if ((suppressed_types_after & conflict) == conflict) {
+			/* Flapping start and end cancel out each other. */
+			suppressed_types_after &= ~conflict;
+		}
 
-			if ((suppressed_types_after & conflict) == conflict) {
-				suppressed_types_after &= ~conflict;
-			}
+		int stateNotifications = NotificationRecovery | NotificationProblem;
+		if (!(suppressed_types_before & stateNotifications) && (suppressed_types & stateNotifications)) {
```

Comment:
OK, stared long enough on this line to give up all this for now. We can continue this together. On a whiteboard.

Reply:
I've extended the PR description (the numbered list on top is new), maybe this helps to understand. This condition is the check if either of the two flags is set for the very first time.

Reply:
And... if none are set?

Reply:
What exactly do you mean? Set as in added to `suppressed_notifications` in general by this function? Or in some of the variables specifically?

Reply:
I mean: to me this looks like intended as XOR, but implemented as NAND – one more trigger case. Or does the latter never happen?

Reply:
Maybe it's easier to read the conditions the other way around:

* `suppressed_types & stateNotifications`: This means that the check result would have triggered a state notification (i.e. either a problem or a recovery notification) if it wasn't suppressed and therefore is stored for later.
* `!(suppressed_types_before & stateNotifications)`: This part checks that neither `NotificationProblem` nor `NotificationRecovery` were set before. So this is the very first time a state notification is generated during the current suppression period.

With the new logic, it isn't really necessary to differentiate between the two types of state notifications. All the code would also work if only a single "NotificationState" flag was added to `suppressed_notifications`. It's mainly done this way as the bitset and `NotificationType` enum already exist in their current form.

For this PR, you could basically ignore states where only one of `NotificationProblem`/`NotificationRecovery` is set and treat them as if both were set. The code in `FireSuppressedNotifications` will figure out which one is appropriate in the end, if either.

Reply:
Ah, now I get it! (!BEFORE) && AFTER